### PR TITLE
Fix OIDC refresh token flow when using the cookie splitter

### PR DESCRIPTION
**Pull request by jochen-kressin**
### Description
This PR fixes a regression regarding the OIDC refresh token flow, caused by the move to the split cookies here: https://github.com/opensearch-project/security-dashboards-plugin/pull/1352

It closes https://github.com/opensearch-project/security-dashboards-plugin/issues/1522 and is based on the research made by @cwperks in https://github.com/opensearch-project/security-dashboards-plugin/pull/1569.

In short, while the refresh token flow was still working correctly, the cookie splitter would not have access to the new idToken that was written to the cookie in the process of refreshing the token. Even though the new value was written, the cookie splitter could not access that value until the next request lifecycle. This caused a "Token expired" and the user would be unauthenticated.

### TODO
End to end testing, but we will most likely do that in a separate PR.

### Category
Bug fix

### Why these changes are required?
To make sure that the OIDC refresh token flow works as expected.

### What is the old behavior before changes and new behavior after changes?
If an idToken expires while the user is idle, the user would be unauthenticated even though the IdP provided a refreshed token.

The new behaviour corrects the refresh token flow, and the user stays authenticated.

### Issues Resolved
https://github.com/opensearch-project/security-dashboards-plugin/issues/1522 

### Testing
Unit tests were added

### Check List
- [x] New functionality includes testing
- [ ] New functionality has been documented N/A
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/session/cookie_splitter.ts [hunk 1]
```diff
@@ -134,12 +143,23 @@ export function unsplitCookiesIntoValue(
...
   additionalCookies: number
 ): string {
-  const rawRequest: HapiRequest = ensureRawRequest(request);
+  const rawRequest: HapiRequestWithStates = ensureRawRequest(request) as HapiRequestWithStates;
   let fullCookieValue = '';
 
+  // We don't want to mix and match between _states and .state.
+  // If we find the first additional cookie in _states, we
+  // use _states for all subsequent additional cookies
+  const requestHasNewerCookieState = rawRequest._states && rawRequest._states[cookiePrefix + 1];
```
Comment (jochen-kressin) [inline server/session/cookie_splitter.ts:152]: @cwperks Here I deviated a bit from the original code that I posted in your PR.

I added a check to make sure that we don't mix and match between whatever is in '_states' and '.state'.
Probably just a theoretical risk, but if the new token would use less cookies than the previous one, the unsplit cookie would contain a mix between the two without this check in place. Hope that's ok.

**Comment by jochen-kressin** [comment]
I removed the Draft status from this PR. While we still need integration tests for this, it probably makes sense to add those in a separate PR even though they should end up in this repository.

So from my side I'd consider this PR finished unless there are any changes wanted. Feedback is always welcome :)

File: server/session/cookie_splitter.ts (hunk 1 above)
Comment (DarshitChanpura) [reply server/session/cookie_splitter.ts:152]: This makes sense to me. We would not want to end up with a bad cookie caused due to mix-up.

**Review by DarshitChanpura — APPROVED** [review]
LGTM!

**Review by cwperks — APPROVED** [review]

**Comment by DarshitChanpura** [comment]
CI  is being fixed by: https://github.com/opensearch-project/security-dashboards-plugin/pull/1596

**Comment by DarshitChanpura** [comment]
> @DarshitChanpura Thanks for merging!
> 
> As per Peter's comment: [#1567 (comment)](https://github.com/opensearch-project/security-dashboards-plugin/issues/1567#issuecomment-1714515170), this should be backported to 1.x together with the "offending" original cookie splitting PR: #1352
> 
> What is the process for getting things into the correct branch...? Should I try to combine the two PRs into one, and then open a PR with 1.x as source and target branch?

IMO they should be combine into one so we have a clean commit history.

**Comment by RyanL1997** [comment]
I have tested @jochen-kressin's security frontend fork with my local set up based on `main` for these [failing cypress tests](https://github.com/opensearch-project/security-dashboards-plugin/actions/runs/6395013515/attempts/7?pr=1580) and everything is all green:
1. `aggregation_view.js`:
Screenshot 2023-10-05 at 1 00 51 PM
2. `multi_tenancy.js`:
Screenshot 2023-10-05 at 12 39 14 PM
3. `default_tenant.js`: 
Screenshot 2023-10-05 at 12 39 31 PM

**Comment by RyanL1997** [comment]
Merging, for more details about the cypress failure lets followup on https://github.com/opensearch-project/security-dashboards-plugin/issues/1599

**Pull request merged**